cds_ils/migrator/series/api.py

```diff
@@ -66,75 +72,109 @@ def create_multipart_volumes(pid, multipart_legacy_recid, migration_volumes):
+            })
+        else:
+            for key in obj:
+                if key != "volume":
+                    if key in volume:
+                        raise KeyError(
+                            'Duplicate key "{}" for multipart {}'.format(
+                                key, multipart_legacy_recid
+                            )
                         )
-                    )
-                volume[key] = obj[key]
+                    volume[key] = obj[key]
 
     volume_numbers = iter(sorted(volumes.keys()))
 
-    # Re-use the current record for the first volume
-    # TODO review this - there are more cases of multiparts
-    first_volume = next(volume_numbers)
-    first = Document.get_record_by_pid(pid)
-    if "title" in volumes[first_volume]:
-        first["title"] = volumes[first_volume]["title"]
-        first["volume"] = first_volume
-    first["_migration"]["multipart_legacy_recid"] = multipart_legacy_recid
+    blueprint = document_blueprint
+    blueprint["_migration"]["multipart_legacy_recid"] = multipart_legacy_recid
+    blueprint["authors"] = blueprint["_migration"]["authors"] \
+        if "authors" in blueprint["_migration"] else []
+    blueprint["serial_title"] = blueprint.get("title")
+
     # to be tested
-    if "legacy_recid" in first:
-        del first["legacy_recid"]
-    first.commit()
-    yield first
+    if "legacy_recid" in blueprint:
+        del blueprint["legacy_recid"]
 
     # Create new records for the rest
     for number in volume_numbers:
-        temp = first.copy()
-        temp["title"] = volumes[number]["title"]
+        volume = volumes[number]
+        temp = blueprint.copy()
+        if "title" in volume and volume["title"]:
+            temp["title"] = volume["title"]
         temp["volume"] = number
         record_uuid = uuid.uuid4()
-        provider = DocumentIdProvider.create(
-            object_type="rec", object_uuid=record_uuid
-        )
-        temp["pid"] = provider.pid.pid_value
-        record = Document.create(temp, record_uuid)
-        record.commit()
-        yield record
+        try:
+            provider = DocumentIdProvider.create(
+                object_type="rec", object_uuid=record_uuid
+            )
+            temp["pid"] = provider.pid.pid_value
+            record = Document.create(temp, record_uuid)
+            record.commit()
+            yield record
+        except IlsValidationError as e:
+            print("Validation error: {}"
+                  .format(str(e.original_exception.message)))
 
 
-def link_and_create_multipart_volumes():
+def link_and_create_multipart_volumes(source):
     """Link and create multipart volume records."""
     click.echo("Creating document volumes and multipart relations...")
-    search = DocumentSearch().filter("term", _migration__is_multipart=True)
-    for hit in search.scan():
-        if "legacy_recid" not in hit:
+    series = {}
+    for hit in SeriesSearch().scan():
```

Comment:
hmm I'm not sure why do you iterate through all the series, instead of searching by the `series_id` of your choice/need. Is it because we handle all the MULTIPARTS-1 and MULTIPARTS-X with the same code?

Reply:
Yes that's the reason. Also I couldn't do a search from the fields inside `_migration` so I had to first fetch all the series and create a local index.

Reply:
What about checking if the field `serial_id` exists?
https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-exists-query.html
This should drastically reduce the number of results, maybe...